dpdispatcher/distributed_shell.py

```diff
@@ -0,0 +1,169 @@
+from dpdispatcher.JobStatus import JobStatus
+from dpdispatcher import dlog
+from dpdispatcher.machine import Machine
+from dpdispatcher.utils import run_cmd_with_all_output
+import subprocess as sp
+
+
+shell_script_header_template="""
+#!/bin/bash -l
+set -x
+"""
+
+script_env_template="""
+{module_unload_part}
+{module_load_part}
+{source_files_part}
+{export_envs_part}
+
+REMOTE_ROOT=`pwd`
+echo 0 > {flag_if_job_task_fail}
+test $? -ne 0 && exit 1
+
+if ! ls {submission_hash}_upload.tgz 1>/dev/null 2>&1; then
+    hadoop fs -get {remote_root}/*.tgz .
+fi
+for TGZ in `ls *.tgz`; do tar xvf $TGZ; done
+
+"""
+script_end_template="""
+cd $REMOTE_ROOT
+test $? -ne 0 && exit 1
+
+wait
+FLAG_IF_JOB_TASK_FAIL=$(cat {flag_if_job_task_fail})
+if test $FLAG_IF_JOB_TASK_FAIL -eq 0; then
+    tar czf {submission_hash}_{job_hash}_download.tar.gz {all_task_dirs}
+    hadoop fs -put -f {submission_hash}_{job_hash}_download.tar.gz {remote_root}
+    hadoop fs -touchz {remote_root}/{job_tag_finished}
+else
+    exit 1
+fi
+"""
+
+class DistributedShell(Machine):
```

Comment:
You may consider to inherit Shell instead considering there are the same methods